feat: use bps instead of `INTERVALS_PER_SLOT` for deadlines

ensi321 opened the pull request:
Closes #8013

ensi321 committed:
Add presets

ensi321 committed:
Update proposer boost

ensi321 committed:
Attestation

ensi321 committed:
Sync committee

ensi321 committed:
Add todo

ensi321 committed:
Block import

ensi321 committed:
lint

ensi321 committed:
update test

ensi321 committed:
fix check-types

ensi321 committed:
Use bps

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

ensi321 committed:
Fix merge

ensi321 committed:
Fix timing

ensi321 committed:
Attempt to fix e2e and sim tests

ensi321 committed:
Fix more

ensi321 committed:
Reduce diff

ensi321 commented on packages/beacon-node/test/e2e/chain/lightclient.test.ts:
```diff
@@ -33,8 +33,9 @@ describe("chain / lightclient", () => {
-  const testParams: Pick<ChainConfig, "SECONDS_PER_SLOT" | "ALTAIR_FORK_EPOCH"> = {
+  const testParams: Pick<ChainConfig, "SECONDS_PER_SLOT" | "SLOT_DURATION_MS" | "ALTAIR_FORK_EPOCH"> = {
     SECONDS_PER_SLOT: 1,
+    SLOT_DURATION_MS: 1000,
```
`SLOT_DURATION_MS` are added to tests like this to make sure the tests run smoothly. 

Officials deprecation of `SECONDS_PER_SLOT` will be done in the future

nflaig replied:
> SLOT_DURATION_MS are added to tests like this to make sure the tests run smoothly

why do we still need `SECONDS_PER_SLOT`? all usage should be removed here but we can keep it in config to avoid interop issues but no longer reference it anywhere in the code

ensi321 replied:
Because a lot of places that are implementation specific still use `SECONDS_PER_SLOT` and I don't want to touch them in this PR. For example `prepareNextSlot` https://github.com/ChainSafe/lodestar/blob/1968319c71046ec0d2526f9387be088b5ac553ac/packages/beacon-node/src/chain/prepareNextSlot.ts#L79

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

nazarhussain commented on packages/state-transition/src/util/slot.ts:
```diff
@@ -21,7 +21,7 @@ export function computeTimeAtSlot(config: ChainConfig, slot: Slot, genesisTime:
-  const timePerInterval = Math.floor(config.SECONDS_PER_SLOT / INTERVALS_PER_SLOT);
-  return Math.floor(secondsIntoSlot / timePerInterval);
+// Convert basis points to milliseconds into the slot
+export function getSlotComponentDuration(config: ChainConfig, basisPoints: number): number {
```
To match the specs. 

```suggestion
export function getSlotComponentDurationMs(config: ChainConfig, basisPoints: number): number {
```

nazarhussain replied:
There is known list of components during the slot, so we can make it more type safe. That will also make code more readable. 

Where instead of calling 

```ts
getSlotComponentDurationMs(config, config.PROPOSER_REORG_CUTOFF_BPS)
```

We would rather call as below, where function knows which components are supported here. Feels like this approach will also make code more readable and easy to know which components we have for slot. 

```
enum SlotComponents {
   PROPOSER_REORG_CUTOFF_BPS: 1, 
    ....
}
getSlotComponentDurationMs(config, SlotComponents.PROPOSER_REORG_CUTOFF_BPS)
```

nazarhussain commented on packages/validator/src/util/params.ts:
```diff
@@ -143,10 +143,17 @@ function getSpecCriticalParams(localConfig: ChainConfig): Record<keyof ConfigWit
 
     // Time parameters
     SECONDS_PER_SLOT: true,
```
Mark it deprecated,

nazarhussain commented on packages/validator/test/unit/services/syncCommittee.test.ts:
```diff
@@ -35,6 +35,7 @@ describe("SyncCommitteeService", () => {
   const config = createChainForkConfig({
     ...mainnetConfig,
     SECONDS_PER_SLOT: 1 / 1000, // Make slot time super short: 1 ms
```
As this one deprecated, we may remove from the tests.

ensi321 replied:
> As this one deprecated, we may remove from the tests.

Deprecation of `SECONDS_PER_SLOT` and properly remove it from our codebase will be done in a separate PR

nazarhussain commented on packages/params/src/index.ts:
```diff
@@ -264,7 +264,8 @@ export const MAX_REQUEST_LIGHT_CLIENT_COMMITTEE_HASHES = 128;
  */
 export const SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY = 128;
-export const INTERVALS_PER_SLOT = 3;
+export const INTERVALS_PER_SLOT = 3; // DEPRECATED
+export const BASIS_POINTS = 10000;
```
Also mark the `SECONDS_PER_SLOT` deprecated.

nazarhussain commented on packages/beacon-node/src/chain/blocks/importBlock.ts:
```diff
@@ -265,8 +265,9 @@ export async function importBlock(
       if (delaySec < SLOTS_PER_EPOCH * this.config.SECONDS_PER_SLOT) {
         this.metrics.importBlock.elapsedTimeTillBecomeHead.observe(delaySec);
-        if (delaySec > this.config.SECONDS_PER_SLOT / INTERVALS_PER_SLOT) {
-          this.metrics.importBlock.setHeadAfterFirstInterval.inc();
+        const cutOffSec = getSlotComponentDuration(this.config, this.config.ATTESTATION_DUE_BPS) / 1000;
```
My preference would be always use `Math.ceil` or `Math.floor` whenever make a division on the `mili seconds` values.

ensi321 replied:
I think we will see `cutOffSec` being something like 3.999999 or 4.000001, so we would need to do `Math.round`. What do you think?

nazarhussain commented on packages/validator/src/services/syncCommittee.ts:
```diff
@@ -270,18 +282,20 @@ export class SyncCommitteeService {
       this.api.validator.submitSyncCommitteeSelections({selections: partialSelections}),
-      // Exit sync committee contributions flow if there is no response after 2/3 of slot.
-      // This is in contrast to attestations aggregations flow which is already exited at 1/3 of the slot
+      // Exit sync committee contributions flow if there is no response after 66% into the slot.
```
Should not we round the decimal of `2/3` into `67%`?

nazarhussain commented on packages/validator/src/util/clock.ts:
```diff
@@ -70,15 +70,22 @@ export class Clock implements IClock {
-  secFromSlot(slot: Slot): number {
-    return Date.now() / 1000 - (this.genesisTime + this.config.SECONDS_PER_SLOT * slot);
+  /** Seconds elapsed from `basisPoints` into a specific slot to now */
+  secFromSlot(slot: Slot, basisPoints = 0): number {
```
Would prefer to either rename this function to properly describe the use of second parameter or have a new function. e.g. `secFromSlotBasePoint` or `secFromSlotComponent`

ensi321 committed:
Address @nazarhussain's comments

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

ensi321 committed:
Address @nazarhussain's comments

ensi321 committed:
Fix calculation

nflaig commented on packages/validator/src/services/attestation.ts:
```diff
@@ -346,12 +355,12 @@ export class AttestationService {
       // due to the fact that calculation of `is_aggregator` in AttestationDutiesService is not done
       // and selectionProof is set to null, meaning no validator will be considered an aggregator.
-      sleep(this.clock.msToSlot(slot + 1 / 3), signal),
+      sleep(this.clock.msToSlotComponent(slot, "AGGREGRATE_DUE_BPS"), signal),
```
`AGGREGRATE_DUE_BPS` is 66% into the slot, not 33%

ensi321 replied:
Good catch

nflaig commented on packages/validator/src/util/clock.ts:
```diff
@@ -76,11 +78,29 @@ export class Clock implements IClock {
+      Date.now() / 1000 -
+      (this.genesisTime +
+        this.config.SECONDS_PER_SLOT * slot +
+        Math.round(getSlotComponentDurationMs(this.config, slotComponent) / 1000))
```
why do we need to round here?

ensi321 replied:
It was suggested here `https://github.com/ChainSafe/lodestar/pull/8091#discussion_r2275705311` . imo I don't think it's that big of a deal to have fraction of a second like 4.0001s or 3.9999s

nflaig replied:
right now it makes not much of a difference as BPS are full second values but if we reduce slot duration it might look different depending on what value we go with and losing accuracy here due to `Math.round` seems unnecessary, see also comment here https://github.com/ChainSafe/lodestar/pull/8091#discussion_r2294067912

ensi321 committed:
Address @nflaig's comment

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

nflaig commented on packages/beacon-node/src/chain/validation/lightClientFinalityUpdate.ts:
```diff
@@ -25,8 +25,10 @@ export function validateLightClientFinalityUpdate(
-  // through the network -- i.e. validate that one-third of finality_update.signature_slot has transpired
-  // (SECONDS_PER_SLOT / INTERVALS_PER_SLOT seconds after the start of the slot, with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+  // through the network --
+  // i.e. validatate that `get_slot_component_duration_ms(SYNC_MESSAGE_DUE_BPS)`
```
new line after `--` is bit strange, same below

nflaig commented on packages/beacon-node/src/chain/validation/lightClientOptimisticUpdate.ts:
```diff
@@ -53,15 +55,19 @@ export function validateLightClientOptimisticUpdate(
 ): boolean {
-  const signatureSlot13TimestampMs = computeTimeAtSlot(config, update.signatureSlot + 1 / 3, genesisTime) * 1000;
-  const earliestAllowedTimestampMs = signatureSlot13TimestampMs - MAXIMUM_GOSSIP_CLOCK_DISPARITY;
+  const updateCutoff =
```
might wanna keep `Ms` suffix
```suggestion
  const updateCutoffMs =
```

nflaig commented on packages/beacon-node/src/chain/validatorMonitor.ts:
```diff
@@ -441,8 +442,12 @@ export function createValidatorMonitor(
+        seenTimestampSec -
+        (genesisTime +
+          data.slot * config.SECONDS_PER_SLOT +
+          Math.round(getSlotComponentDurationMs(config, "ATTESTATION_DUE_BPS") / 1000));
```
I really think we shouldn't use `Math.round`, we already limit the decimals we print below `delaySec.toFixed(4)`

nflaig replied:
we don't round the `seenTimestampSec` either https://github.com/ChainSafe/lodestar/blob/4641b565d55a6ea1982c083fefebf7c71be9ebb9/packages/beacon-node/src/api/impl/beacon/blocks/index.ts#L87

nflaig commented on packages/config/src/chainConfig/types.ts:
```diff
@@ -49,11 +49,18 @@ export type ChainConfig = {
 
   // Time parameters
-  SECONDS_PER_SLOT: number;
+  SECONDS_PER_SLOT: number; // Deprecated
```
might be better to do

```suggestion
  /** @deprecated */
  SECONDS_PER_SLOT: number;
```

nflaig commented on packages/params/src/types.ts:
```diff
@@ -57,7 +57,6 @@ export type BeaconPreset = {
   PROPORTIONAL_SLASHING_MULTIPLIER_ALTAIR: number;
   MIN_SYNC_COMMITTEE_PARTICIPANTS: number;
   UPDATE_TIMEOUT: number;
-
```
:eyes:

nflaig commented on packages/state-transition/src/util/slot.ts:
```diff
@@ -21,7 +21,8 @@ export function computeTimeAtSlot(config: ChainConfig, slot: Slot, genesisTime:
+// Convert basis points to milliseconds into the slot
+export function getSlotComponentDurationMs(config: ChainConfig, slotComponent: SlotComponent): number {
+  const basisPoints = config[slotComponent];
+  return Math.floor((basisPoints * config.SLOT_DURATION_MS) / BASIS_POINTS);
```
why do we need to resolve the basis points inside the function?

can we just call it like this `getSlotComponentDurationMs(config.ATTESTATION_DUE_BPS)`?

ensi321 replied:
Where do we get `config.SLOT_DURATION_MS` from?

You mean we should 
`getSlotComponentDurationMs(config.SLOT_DURATION_MS, config.ATTESTATION_DUE_BPS)`?

nflaig replied:
> getSlotComponentDurationMs(config.SLOT_DURATION_MS, config.ATTESTATION_DUE_BPS)?

`getSlotComponentDurationMs(config, config.ATTESTATION_DUE_BPS)?` or make it part of config object

nflaig replied:
`config.getSlotComponentDurationMs(config.ATTESTATION_DUE_BPS)` would align more closely with spec and I kinda don't like passing these string values as we do right now, wdyt?

ensi321 replied:
> `config.getSlotComponentDurationMs(config.ATTESTATION_DUE_BPS)` would align more closely with spec and I kinda don't like passing these string values as we do right now, wdyt?

I am going to go for 
`function getSlotComponentDurationMs(config: ChainConfig, basisPoints: number): number`
I think it serves a similar purpose to `computeTimeAtSlot`, so similar signature would be good.


Note I am keeping
```
secFromSlotComponent(slot: Slot, slotComponent: SlotComponent): number
msToSlotComponent(slot: Slot, slotComponent: SlotComponent): number 
```
on the vc side. 

Do you think we should make them
```
secFromSlotComponent(slot: Slot, basisPoints: number): number
msToSlotComponent(slot: Slot, basisPoints: number): number 
```
too?

nflaig replied:
having `SlotComponent` is another level of indirection I don't really like that much, I would just get rid of the type for now, it's simple to change this later if we see disadvantages of passing basis points directly from config values

nflaig commented on packages/validator/src/services/syncCommittee.ts:
```diff
@@ -270,18 +282,20 @@ export class SyncCommitteeService {
     if (!res) {
-      throw new Error("Failed to receive combined selection proofs before 2/3 of slot");
+      throw new Error(
+        `Failed to receive combined selection proofs before ${this.config.CONTRIBUTION_DUE_BPS / BASIS_POINTS}% into slot`
```
hmm why do we need to calculate this? `CONTRIBUTION_DUE_BPS` will always be 2/3 of slot

nflaig commented on packages/validator/src/services/attestation.ts:
```diff
@@ -346,12 +355,12 @@ export class AttestationService {
     const res = await Promise.race([
       this.api.validator.submitBeaconCommitteeSelections({selections: partialSelections}),
-      // Exit attestation aggregation flow if there is no response after 1/3 of slot as
+      // Exit attestation aggregation flow if there is no response after 33% into the slot as
```
consider also updating error message on L367 to be consistent with sync committee service

nflaig reviewed (commented):
Did another more detailed review, I think we should get rid of `Math.round`

nflaig commented on packages/config/src/chainConfig/types.ts:
```diff
@@ -49,11 +49,18 @@ export type ChainConfig = {
   ETH1_FOLLOW_DISTANCE: number;
+  PROPOSER_REORG_CUTOFF_BPS: number;
+  ATTESTATION_DUE_BPS: number;
+  AGGREGRATE_DUE_BPS: number;
```
There's a typo here, that's already being fixed in https://github.com/ethereum/consensus-specs/pull/4524 but it would be good if we wait for spec release and update our config tests to make sure all constants match with the spec

```suggestion
  AGGREGATE_DUE_BPS: number;
```

ensi321 replied:
🤦 can't believe I didn't catch this either

ensi321 committed:
Remove Math.round

ensi321 committed:
Update deprecated style

ensi321 committed:
Address various @nflaig's comments

ensi321 committed:
lint

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

ensi321 committed:
Fix typo

ensi321 committed:
Line break in comment

nflaig commented on packages/validator/src/util/params.ts:
```diff
@@ -146,11 +146,19 @@ function getSpecCriticalParams(localConfig: ChainConfig): Record<keyof ConfigWit
     GLOAS_FORK_EPOCH: gloasForkRelevant,
 
     // Time parameters
+    /** @deprecated */
```
this only makes sense on the type definition

<img width="462" height="60" alt="image" src="https://github.com/user-attachments/assets/def648b0-6951-4f26-a669-df73e035123d" />


but not here

ensi321 committed:
Update getSlotComponentDurationMs's signature

ensi321 committed:
Update deprecated style

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

ensi321 commented:
Waiting for ethereum/consensus-specs#4542 to be merged and new release of the spec. Justin is also on vacation so it might take some time

nflaig commented:
@ensi321 I like the suggestion in https://github.com/ethereum/consensus-specs/issues/4586, this would make it pretty similar to how we handle blobs via `getMaxBlobsPerBlock(epoch)`.

I would suggest we go with that approach since we will already need 2 different cutoff for current vs. ePBS and then with slot reduction the intervals will change again, so preparing for that having a getter function is good imo and most explicit.

ensi321 commented:
> @ensi321 I like the suggestion in [ethereum/consensus-specs#4586](https://github.com/ethereum/consensus-specs/issues/4586), this would make it pretty similar to how we handle blobs via `getMaxBlobsPerBlock(epoch)`.
> 
> I would suggest we go with that approach since we will already need 2 different cutoff for current vs. ePBS and then with slot reduction the intervals will change again, so preparing for that having a getter function is good imo and most explicit.

Yes, I think the spec is going that direction. Are you suggesting we should implement this regardless of how the spec will be written?

nflaig commented:
> Are you suggesting we should implement this regardless of how the spec will be written?

I would just go ahead and implement it, ie. add helpers like `getAttestationDueMs(epoch)` etc. This will help working on glamsterdam stuff

nflaig commented:
looks like https://github.com/ethereum/consensus-specs/pull/4589 will be merged as well

ensi321 committed:
Update vc

ensi321 committed:
Update bn

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

nflaig commented on packages/state-transition/src/util/slot.ts:
```diff
@@ -21,7 +21,27 @@ export function computeTimeAtSlot(config: ChainConfig, slot: Slot, genesisTime:
-export function getCurrentInterval(config: ChainConfig, secondsIntoSlot: number): number {
-  const timePerInterval = Math.floor(config.SECONDS_PER_SLOT / INTERVALS_PER_SLOT);
-  return Math.floor(secondsIntoSlot / timePerInterval);
+export function getAttestationDueMs(config: ChainConfig): number {
```
can we already pass epoch here, this reduces the diff later on for ePBS

```suggestion
export function getAttestationDueMs(config: ChainConfig, _epoch: Epoch): number {
```

ensi321 replied:
I think we can do `_fork: ForkName` instead, since changes deadlines requiring a new hard fork

nflaig replied:
good point, makes sense to do that instead

nflaig commented on packages/state-transition/src/util/slot.ts:
```diff
@@ -21,7 +21,27 @@ export function computeTimeAtSlot(config: ChainConfig, slot: Slot, genesisTime:
-  const timePerInterval = Math.floor(config.SECONDS_PER_SLOT / INTERVALS_PER_SLOT);
-  return Math.floor(secondsIntoSlot / timePerInterval);
+export function getAttestationDueMs(config: ChainConfig): number {
+  return getSlotComponentDurationMs(config, config.ATTESTATION_DUE_BPS);
```
we can keep them here but I slightly would prefer them to be part of config object, same as `getMaxBlobsPerBlock`


also noticed we probably wanna get rid of `getMaxRequestBlobSidecars` in config object as it's only pre-fulu

ensi321 replied:
> we can keep them here but I slightly would prefer them to be part of config object, same as `getMaxBlobsPerBlock`

I agree. Especially when the due times are changed in gloas and one more time when we get to 6s slot

> also noticed we probably wanna get rid of getMaxRequestBlobSidecars in config object as it's only pre-fulu

will have a separate PR for it

nflaig commented on packages/beacon-node/src/metrics/metrics/lodestar.ts:
```diff
@@ -796,9 +796,9 @@ export function createLodestarMetrics(
-        help: "Total times an imported block is set as head after the first slot interval",
+      setHeadAfterCutoff: register.gauge({
+        name: "lodestar_import_block_set_head_after_cutoff_total",
+        help: "Total times an imported block is set as head after 33% of the slot",
```
just realized we should avoid using % in comments and logs as it will be inaccurate with ePBS

nflaig commented on packages/validator/src/metrics.ts:
```diff
@@ -82,25 +82,25 @@ export function getMetrics(register: MetricsRegisterExtra, gitData: LodestarGitD
-      help: "Time between 1/3 of slot and call produce message",
-      // Max wait time is 1 / 3 of slot
+      help: "Time between SYNC_MESSAGE_DUE_BPS of slot and call produce message",
+      // Max wait time is 4s of slot
```
same here, probably wanna do 
```suggestion
      // Max wait time is SYNC_MESSAGE_DUE_BPS of slot
```

nflaig commented on packages/validator/src/services/syncCommittee.ts:
```diff
@@ -83,19 +84,19 @@ export class SyncCommitteeService {
-      await sleep(this.clock.msToSlot(slot + 2 / 3), signal);
-      this.metrics?.syncCommitteeStepCallProduceAggregate.observe(this.clock.secFromSlot(slot + 2 / 3));
+      // First, wait until the `CONTRIBUTION_DUE_BPS` (67% into the slot)
+      await sleep(this.clock.secFromSyncContributionDue(slot), signal);
```
`secFromSyncContributionDue` returns seconds but `sleep` expects milliseconds

nflaig commented on packages/validator/src/services/syncCommittee.ts:
```diff
@@ -83,19 +84,19 @@ export class SyncCommitteeService {
       // see https://github.com/ChainSafe/lodestar/issues/4608
-      await Promise.race([sleep(this.clock.msToSlot(slot + 1 / 3), signal), this.emitter.waitForBlockSlot(slot)]);
-      this.metrics?.syncCommitteeStepCallProduceMessage.observe(this.clock.secFromSlot(slot + 1 / 3));
+      await Promise.race([sleep(this.clock.secFromSyncMessageDue(slot), signal), this.emitter.waitForBlockSlot(slot)]);
```
same as code below, need to pass milliseconds to `sleep`

nflaig commented on packages/validator/src/util/clock.ts:
```diff
@@ -81,6 +82,39 @@ export class Clock implements IClock {
     return Date.now() / 1000 - (this.genesisTime + this.config.SECONDS_PER_SLOT * slot);
   }
 
+  msToAttestationDue(slot: number): number {
```
do we really need those? need to maintain them separately as they are no reusing other helpers

there also doesn't seem to be a reason to calculate slot fraction, eg. in attestation service can just do
```ts
await Promise.race([sleep(this.clock.msToSlot(slot) + getAttestationDueMs(this.config), signal), this.emitter.waitForBlockSlot(slot)]);
```

or we add a `msFromSlot` helper but it's just semantics

ensi321 committed:
Fix sync committee function calling

ensi321 committed:
Move timing functions to inside fork config

ensi321 committed:
Update comment

ensi321 committed:
Remove ISlotComponentClock

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

ensi321 committed:
lint

ensi321 committed:
Fix check-types

nflaig commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -254,7 +260,8 @@ export class BeaconChain implements IBeaconChain {
 
     this.blacklistedBlocks = new Map((opts.blacklistedBlocks ?? []).map((hex) => [hex, null]));
-    const preAggregateCutOffTime = (2 / 3) * this.config.SECONDS_PER_SLOT;
+    // TODO GLOAS: Pass in a real fork name
```
why not pass in real fork already?

ensi321 replied:
Some of the callers don't have direct access to fork/slot/epoch info, may potentially need some refactoring. So delegate it to another PR

nflaig replied:
we should at least pass the real fork already where it's trivial, adding a TODO in those places doesn't make sense

> So delegate it to another PR

we should do any kind of bigger refactoring now, this PR should make our code base work properly with different slot intervals, otherwise it's just incomplete imo

nflaig replied:
I resolved all of the `TODO GLOAS: Pass in a real fork name` comments, was trivial and the goal of this PR should be imo that we prepare this as much as possible to work with gloas

nflaig commented on packages/config/src/forkConfig/types.ts:
```diff
@@ -56,4 +56,12 @@ export type ForkConfig = {
+  getSyncContributionDueMs(_fork: ForkName): number;
+  getProposerReorgCutoffMs(_fork: ForkName): number;
+  // Convert basis points to milliseconds into the slot
+  getSlotComponentDurationMs(basisPoints: number): number;
```
this seems like an internal helper, should not be exposing that as part of config object

ensi321 committed:
Address @nflaig's comment

nflaig reviewed (commented):
Just noticed we might wanna set `SLOT_DURATION_MS` if someone configures `SECONDS_PER_SLOT` via kurtosis (or similar) to keep backward compatibility

ensi321 commented:
> Just noticed we might wanna set `SLOT_DURATION_MS` if someone configures `SECONDS_PER_SLOT` via kurtosis (or similar) to keep backward compatibility

I guess we can provide this functionality until Gloas is live. After that we expect `SECONDS_PER_SLOT` to be fully deprecated and is a no-op when setting it

ensi321 committed:
Set `SLOT_DURATION_MS` if not provided

ensi321 committed:
Merge branch 'unstable' into nc/timing-refactor

nflaig committed:
Use getAggregateDueMs instead of 2/3

nflaig committed:
Remove skipped constants from config test

nflaig commented on packages/validator/src/services/attestation.ts:
```diff
@@ -88,9 +88,15 @@ export class AttestationService {
+      this.emitter.waitForBlockSlot(slot),
+    ]);
+    this.metrics?.attesterStepCallProduceAttestation.observe(
+      this.clock.secFromSlot(slot) + this.config.getAttestationDueMs(ForkName.phase0) / 1000
```
this doesn't seem right, need to subtract the deadline from this value as it supposed to track time between cutoff and call
- so if sec from slot is 2, this should be -2
- if sec from slot is 4, this should be 0

these are our current metrics which confirm this, mostly negative values because we receive block earlier than cutoff

<img width="938" height="289" alt="image" src="https://github.com/user-attachments/assets/dffdebd9-a456-4793-8fed-6e57d4e5e3a5" />

nflaig replied:
this should be resolved, need to double check on `feat4` once there are some metrics

nflaig replied:
this is from `feat4` and looks as expected now

<img width="943" height="278" alt="image" src="https://github.com/user-attachments/assets/df8a2d42-6be2-4460-9eb3-550039f703a7" />

nflaig commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -254,7 +260,8 @@ export class BeaconChain implements IBeaconChain {
     this.blacklistedBlocks = new Map((opts.blacklistedBlocks ?? []).map((hex) => [hex, null]));
-    const preAggregateCutOffTime = (2 / 3) * this.config.SECONDS_PER_SLOT;
+    // TODO GLOAS: Pass in a real fork name
+    const preAggregateCutOffTime = config.getAggregateDueMs(ForkName.phase0) / 1000;
```
This is not right, first it passes same cutoff to sync committee pool which is likely not a big problem but also it initializes the value in constructor but we need to handle different cutoffs between fork and this here would only get updated once you restart the node.

nflaig replied:
will be resolved by https://github.com/ChainSafe/lodestar/pull/8459

nflaig committed:
Add todo

nflaig commented on packages/beacon-node/src/chain/validation/lightClientOptimisticUpdate.ts:
```diff
@@ -26,8 +27,9 @@ export function validateLightClientOptimisticUpdate(
   // [IGNORE] The optimistic_update is received after the block at signature_slot was given enough time to propagate
-  // through the network -- i.e. validate that one-third of optimistic_update.signature_slot has transpired
-  // (SECONDS_PER_SLOT / INTERVALS_PER_SLOT seconds after the start of the slot, with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+  // through the network -- i.e. validate that `get_slot_component_duration_ms(SYNC_MESSAGE_DUE_BPS)`
```
all of these comments are not longer matching spec since it was updated in https://github.com/ethereum/consensus-specs/pull/4589

nflaig replied:
resolved

nflaig committed:
Fix collecting vc timings

nflaig committed:
Properly deprecate

nflaig committed:
More comment updates

nflaig committed:
Update vc dashboard titles

nflaig committed:
Update comments to match CL spec

nflaig commented on packages/beacon-node/src/sync/unknownBlock.ts:
```diff
@@ -92,7 +92,8 @@ export class BlockInputSync {
     this.maxPendingBlocks = opts?.maxPendingBlocks ?? MAX_PENDING_BLOCKS;
-    this.proposerBoostSecWindow = this.config.SECONDS_PER_SLOT / INTERVALS_PER_SLOT;
+    // TODO GLOAS: Pass in a real fork name
+    this.proposerBoostSecWindow = config.getAttestationDueMs(ForkName.phase0) / 1000;
```
same here, it's not correct to set this in constructor anymore as we need those timings to be dynamic

nflaig replied:
will be resolved by https://github.com/ChainSafe/lodestar/pull/8459

nflaig committed:
Pass actual fork

nflaig committed:
Add todo

nflaig committed:
Update more spec references

nflaig committed:
Fix lint/type checks

nflaig committed:
fix: make cutoffs initialized in constructors dynamic (#8459)

Addresses
https://github.com/ChainSafe/lodestar/pull/8091#discussion_r2375439397
and
https://github.com/ChainSafe/lodestar/pull/8091#discussion_r2375272330

nflaig committed:
Misc changes

nflaig committed:
Fix interface

nflaig committed:
Deduplicate stuff

nflaig committed:
Update todo

nflaig commented on packages/beacon-node/src/chain/validatorMonitor.ts:
```diff
@@ -34,6 +34,7 @@ const MAX_CACHED_EPOCHS = 4;
 
 const MAX_CACHED_DISTINCT_TARGETS = 4;
 
+// TODO GLOAS: update to use BPS and re-evaluate values
```
we might already wanna do this as part of the PR, there really shouldn't be any remaining usage of `INTERVALS_PER_SLOT`

nflaig replied:
will be addressed by https://github.com/ChainSafe/lodestar/pull/8471

nflaig commented on packages/validator/src/services/attestation.ts:
need to review usage of `msToSlot`, there are still a few places that use  `1 / 3`
https://github.com/ChainSafe/lodestar/blob/83de5b8deac20e6b969314899644ecb4e6df3775/packages/validator/src/services/attestationDuties.ts#L137-L138

nflaig replied:
added a todo, should be fine for now

nflaig committed:
Simplify waitForSyncMessageCutoff

nflaig committed:
Remove stray value

nflaig committed:
refactor: add msFromSlot to validator clock (#8463)

This makes it easier to reason about the code

nflaig committed:
Simplify updateReceivedTooEarly

nflaig commented:
/gemini review

nflaig commented on packages/beacon-node/src/api/impl/validator/index.ts:
```diff
@@ -108,6 +108,8 @@ export const SYNC_TOLERANCE_EPOCHS = 1;
+ *
+ * TODO GLOAS: re-evaluate cutoff timing
  */
 const BLOCK_PRODUCTION_RACE_CUTOFF_MS = 2_000;
```
@gemini-code-assist there are not the same, `BLOCK_PRODUCTION_RACE_CUTOFF_MS` is a custom cutoff not defined in the spec

nflaig committed:
Fix light client unit tests

nflaig committed:
Use Math.round

nflaig commented on packages/config/src/forkConfig/index.ts:
```diff
@@ -117,7 +117,7 @@ export function createForkConfig(config: ChainConfig): ForkConfig {
   // Convert basis points to milliseconds into the slot
   const getSlotComponentDurationMs = (basisPoints: number): number =>
-    Math.floor((basisPoints * config.SLOT_DURATION_MS) / BASIS_POINTS);
+    Math.round((basisPoints * config.SLOT_DURATION_MS) / BASIS_POINTS);
```
I think this is cleaner, noticed attestation deadline is 3999ms otherwise, not as per spec but doesn't matter in this case, the difference is negligible

ensi321 replied:
Yea worst case is +/- 0.5ms, which is negligible

nflaig committed:
Merge branch 'unstable' into nc/timing-refactor

wemeetagain commented on packages/beacon-node/src/chain/validatorMonitor.ts:
```diff
@@ -450,8 +451,11 @@ export function createValidatorMonitor(
+      // Returns the duration between when the attestation `data` could be produced (ATTESTATION_DUE_BPS through the slot) and `seenTimestamp`.
+      const delaySec =
+        seenTimestampSec -
+        (genesisTime + data.slot * config.SECONDS_PER_SLOT + config.getAttestationDueMs(fork) / 1000);
```
should this be refactored to not use SECONDS_PER_SLOT?
I suppose we would need something like config.getTimeAtSlot(slot) or similar

nflaig replied:
I was in favor of removing usage of `SECONDS_PER_SLOT` in this PR too but @ensi321 didn't want it, we can do it in a follow up and simplify this code a bit

ensi321 replied:
> I was in favor of removing usage of `SECONDS_PER_SLOT` in this PR too but @ensi321 didn't want it, we can do it in a follow up and simplify this code a bit

It's fine to do in this PR or next PR. Was trying to limit the scope of this PR

nflaig committed:
Add todo to re-evaluate 1/3 sleep

nflaig committed:
Merge branch 'unstable' into nc/timing-refactor

nflaig commented on packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:
```diff
@@ -464,6 +464,7 @@ export class PersistentCheckpointStateCache implements CheckpointStateCache {
 
     const blockSlot = state.slot;
+    // TODO GLOAS: this needs to be re-evaluated, should it be based on AGGREGATE_DUE_BPS?
     const twoThirdsSlot = (2 * state.config.SECONDS_PER_SLOT) / INTERVALS_PER_SLOT;
```
will be addressed by https://github.com/ChainSafe/lodestar/pull/8471

nflaig committed:
chore: remove remaining usage of intervals (#8471)

nflaig commented:
/gemini review

nflaig commented on packages/validator/src/util/params.ts:
```diff
@@ -146,11 +146,18 @@ function getSpecCriticalParams(localConfig: ChainConfig): Record<keyof ConfigWit
 
     // Time parameters
-    SECONDS_PER_SLOT: true,
+    SECONDS_PER_SLOT: true, // Deprecated
```
we could consider to stop validating this once we have removed all usage of `SECONDS_PER_SLOT`

nflaig replied:
will be addressed by https://github.com/ChainSafe/lodestar/pull/8476

ensi321 commented on packages/beacon-node/src/chain/validatorMonitor.ts:
```diff
@@ -492,11 +487,11 @@ export function createValidatorMonitor(
       const src = OpSource.gossip;
       const data = indexedAttestation.data;
       const epoch = computeEpochAtSlot(data.slot);
+      const fork = config.getForkName(data.slot);
```
Determining hard fork based on `data.slot` could potentially have issue at fork boundary. 

So when we switch over to 6s slot, we will need to revisit all these timings.

nflaig replied:
In this case it's what we want since we evaluate the attestation based on the fork it belongs to, we also use `data.slot` in the calculation to determine the delay.

wemeetagain commented:
:tada: This PR is included in [v1.35.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.35.0) :tada:
